add FreeBSD IP_RECVDSTADDR support

lucifer9 opened the pull request:
Fixes #1142

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -230,8 +262,18 @@ fn send(
     let num_transmits = transmits.len().min(BATCH_SIZE);
 
     loop {
+        #[cfg(not(target_os = "freebsd"))]
         let n =
             unsafe { libc::sendmmsg(io.as_raw_fd(), msgs.as_mut_ptr(), num_transmits as u32, 0) };
+        #[cfg(target_os = "freebsd")]
+        let n = unsafe {
+            libc::sendmmsg(
+                io.as_raw_fd(),
+                msgs.as_mut_ptr(),
+                num_transmits as libc::size_t,
+                0,
+            )
+        };
```
Can we fold these together by using `as _`, or at least isolate the cfg to declaring an intermediate variable? It took me a solid thirty seconds to find the difference between these two lines.

lucifer9 replied:
You are right, let's use ```as _```.

Ralith replied:
Looks like this hasn't been done yet.

lucifer9 replied:
Sorry, I reverted too much last time.

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -427,7 +480,8 @@ fn prepare_msg(
 
     if let Some(ip) = &transmit.src_ip {
-        if cfg!(target_os = "linux") {
+        #[cfg(target_os = "linux")]
```
No way to do this on FreeBSD?

lucifer9 replied:
I have tried the similar operations on FreeBSD, but failed on most tests.
BTW, I'm not quite sure what this piece of code does.

Ralith replied:
This sets the source IP address of an outgoing packet, which allows quinn to be well-behaved when an interface has multiple addresses on the same network. It's not required, but it's nice to have when possible.

lucifer9 replied:
The IPv6 part works on FreeBSD with no change.
I tried to modify IPv4 part as below
```
                #[cfg(target_os = "freebsd")]
                IpAddr::V4(v4) => {
                    let addr = libc::in_addr {
                        s_addr: u32::from_ne_bytes(v4.octets()),
                    };
                    encoder.push(libc::IPPROTO_IP, libc::IP_SENDSRCADDR, addr);
                }
```
But some tests failed, please see the [log](https://github.com/quinn-rs/quinn/files/10088476/log.txt)

Ralith replied:
Seems like that must not be the right way to do it, then. We can leave this for future work.

lucifer9 replied:
I know what's wrong: sendmsg will fail with EINVAL if IP_SENDSRCADDR is used with a socket which is bound to something other than Ipv4Addr::UNSPECIFIED.
I added a check to make sure encode source ip in cmsg only when bind to Ipv4Addr::UNSPECIFIED.

Ralith replied:
Hmm, what if the interface has multiple addresses bound to it? Does FreeBSD not support that?

lucifer9 replied:
FreeBSD supports multiple IPs on a single interface.
But when a socket does not bind to Ipv4Addr::UNSPECIFIED (using bind()), FreeBSD's sendmsg just fails with EINVAL.
In the tests, eg. echo_v4, client binds to Ipv4Addr::UNSPECIFIED and server binds to Ipv4Addr::LOCALHOST, so client sendmsg works well but server fails with `sendmsg error: Os { code: 22, kind: InvalidInput, message: "Invalid argument" }`

Ralith replied:
>  But when a socket does not bind to Ipv4Addr::UNSPECIFIED (using bind()), FreeBSD's sendmsg just fails with EINVAL.

Right, so on the surface this seems to prevent selecting the right IP address to respond from, if you have more than one address on the same subnet, which is weird. Not a blocker here, though, this is an improvement regardless.

Ralith replied:
Nevermind, I confused myself, this behavior makes sense. Obviously if a socket is bound to a specific address you only need to send from that address.

Ralith reviewed (commented):
Thanks! Any chance you could add a FreeBSD case to CI so this doesn't bitrot?

lucifer9 commented:
There's no FreeBSD runner image. So I try to use [vmactions/freebsd-vm](https://github.com/vmactions/freebsd-vm) to run tests.

Ralith commented on .github/workflows/rust.yml:
```diff
@@ -8,6 +8,28 @@ on:
+          usesh: true
+          mem: 4096
+          copyback: false
+          prepare: |
+            pkg install -y curl
+            curl https://sh.rustup.rs -sSf --output rustup.sh
+            sh rustup.sh -y --profile minimal --default-toolchain stable
+            echo "~~~~ rustc --version ~~~~"
+            $HOME/.cargo/bin/rustc --version
+          run: |
+            freebsd-version
+            $HOME/.cargo/bin/cargo build --all-targets
+            $HOME/.cargo/bin/cargo test
+            $HOME/.cargo/bin/cargo test --manifest-path fuzz/Cargo.toml
```
Do the actions-rs helpers not work here? If so, it'd be nice to leave a comment citing an upstream issue so that can be fixed eventually.

lucifer9 replied:
I'm afraid not. Github actions have no support for FreeBSD https://github.com/actions/runner/issues/385
I'll leave a comment as you said.

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -180,6 +180,38 @@ fn init(io: SockRef<'_>) -> io::Result<()> {
+                return Err(io::Error::last_os_error());
+            }
+        } else {
+            let on: libc::c_int = 1;
+            let rc = unsafe {
+                libc::setsockopt(
+                    io.as_raw_fd(),
+                    libc::IPPROTO_IPV6,
+                    libc::IPV6_RECVPKTINFO,
+                    &on as *const _ as _,
+                    mem::size_of_val(&on) as _,
+                )
+            };
+            if rc == -1 {
+                return Err(io::Error::last_os_error());
+            }
```
Could we move this down into the common `is_ipv4` case below, and delete the identical Linux `IPV6_RECVPKTINFO` call above?

lucifer9 replied:
Sure.

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -196,7 +182,12 @@ fn init(io: SockRef<'_>) -> io::Result<()> {
+    }
+
+    if !is_ipv4 {
+        #[cfg(any(target_os = "linux", target_os = "freebsd"))]
```
Is there any reason to believe this is needed? `IPV6_RECVPKTINFO` is standardized.

lucifer9 replied:
Yes, IPV6_RECVPKTINFO is standardized. 
But for other OS, eg. macOS, it will fail test echo_v6 because in quinn/src/tests.rs we do not distinguish between ipv4 and ipv6.
So I tried to add the ipv4 (IP_RECVDSTADDR) part to macOS, too. 
BTW, It is interesting that macOS supports both IP_RECVDSTADDR & IP_PKTINFO. 
It seems iOS will also work but I have no devices to test. ( I have MacBook Air & a Linux PVE server which runs a FreeBSD vm.)

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -432,17 +466,29 @@ fn prepare_msg(
 
     if let Some(ip) = &transmit.src_ip {
-        if cfg!(target_os = "linux") {
+        {
```
Unnecessary block?

lucifer9 replied:
You are right. I'll remove it.

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -444,16 +483,26 @@ fn prepare_msg(
-                        },
-                    };
-                    encoder.push(libc::IPPROTO_IPV6, libc::IPV6_PKTINFO, pktinfo);
+                #[cfg(any(target_os = "freebsd", target_os = "macos"))]
+                {
+                    if encode_srcip {
```
It looks like the macOS version of `send` always passes `false` for `encode_srcip`, rendering this code dead on macOS, even though the cfg gate ensures it's compiled. Is that intended?

lucifer9 replied:
Sorry, corrected now.

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -444,16 +483,26 @@ fn prepare_msg(
+                        let addr = libc::in_addr {
+                            s_addr: u32::from_ne_bytes(v4.octets()),
+                        };
+                        dbg!(addr.s_addr);
```
Stray dbg

lucifer9 replied:
Removed.

Ralith commented on quinn-udp/src/unix.rs:
```diff
@@ -330,7 +339,7 @@ fn send(
             &mut iov,
             &mut ctrl,
-            false,
+            encode_srcip,
```
nit: pass `cfg!(target_os = "macos")` directly here and we don't need to bother defining and mutating a separate `encode_srcip` variable.

djc reviewed (approved):
Thank you for fixing up our FreeBSD support!

I had a bit of a hard  time reviewing this because it seems to have grown to encompass a bit more than just `RECVDSTADDR`. If I understand correctly, there are few things here:

* Set `IPV6_RECVPKTINFO` on FreeBSD as well (not just on Linux)
* Set `IP_RECVDSTADDR` on FreeBSD and macOS only for IPv4
* Fix a warning about unused `state: &UdpState` which we didn't notice before
* Allow different types for the 3rd argument to `sendmmsg()`/`recvmmsg()`

It might be nice add some comments about how FreeBSD compares to Linux here, which is left implicit, for the reader to decode from the code. Additionally, IMO the expected spelling for `encode_srcip` would be `encode_src_ip` -- might be nice to fix that up.

lucifer9 commented:
> Thank you for fixing up our FreeBSD support (and adding a CI test for it)!
> 
> I had a bit of a hard time reviewing this because it seems to have grown to encompass a bit more than just `RECVDSTADDR`. If I understand correctly, there are few things here:
> 
> * Set `IPV6_RECVPKTINFO` on FreeBSD as well (not just on Linux)
> * Set `IP_RECVDSTADDR` on FreeBSD and macOS only for IPv4
> * Fix a warning about unused `state: &UdpState` which we didn't notice before
> * Allow different types for the 3rd argument to `sendmmsg()`/`recvmmsg()`
> 
> It might be nice add some comments about how FreeBSD compares to Linux here, which is left implicit, for the reader to decode from the code. Additionally, IMO the expected spelling for `encode_srcip` would be `encode_src_ip` -- might be nice to fix that up.

I added  a few comments to clarify these parts related to FreeBSD and macOS, hope they be useful. And I have fixed the spelling😊

asomers commented:
BTW, the easiest way to do CI on FreeBSD for Github-hosted projects is with https://cirrus-ci.com/ .

littlesum commented:
> Fixes #1142

I see you fix this bug in freebsd, but it still not work in openbsd, is there any possible help fix it in openbsd

```
sur# cargo build
    Updating crates.io index
  Downloaded matchers v0.1.0
  Downloaded num_threads v0.1.6
  Downloaded heck v0.4.0
  Downloaded thread_local v1.1.4
  Downloaded nu-ansi-term v0.46.0
  Downloaded bytes v1.3.0
  Downloaded anyhow v1.0.66
  Downloaded tracing-subscriber v0.3.16
  Downloaded hdrhistogram v7.5.2
  Downloaded tokio v1.23.0
  Downloaded time-core v0.1.0
  Downloaded rustc-hash v1.1.0
  Downloaded termcolor v1.1.3
  Downloaded proc-macro-error v1.0.4
  Downloaded strsim v0.10.0
  Downloaded time-macros v0.2.6
  Downloaded yasna v0.5.0
  Downloaded time v0.3.17
  Downloaded clap_derive v3.2.18
  Downloaded sharded-slab v0.1.4
  Downloaded signal-hook-registry v1.4.0
  Downloaded overload v0.1.1
  Downloaded rcgen v0.10.0
  Downloaded proc-macro-error-attr v1.0.4
  Downloaded pem v1.1.0
  Downloaded 25 crates (1.7 MB) in 5.49s
   Compiling proc-macro2 v1.0.47
   Compiling unicode-ident v1.0.5
   Compiling quote v1.0.21
   Compiling libc v0.2.138
   Compiling syn v1.0.105
   Compiling autocfg v1.1.0
   Compiling once_cell v1.16.0
   Compiling cfg-if v1.0.0
   Compiling cc v1.0.77
   Compiling untrusted v0.7.1
   Compiling spin v0.5.2
   Compiling version_check v0.9.4
   Compiling base64 v0.13.1
   Compiling pin-project-lite v0.2.9
   Compiling itoa v1.0.4
   Compiling log v0.4.17
   Compiling rustls v0.20.7
   Compiling ppv-lite86 v0.2.17
   Compiling thiserror v1.0.37
   Compiling openssl-probe v0.1.5
   Compiling regex-syntax v0.6.28
   Compiling tinyvec_macros v0.1.0
   Compiling time-core v0.1.0
   Compiling num_threads v0.1.6
   Compiling rustc-hash v1.1.0
   Compiling bytes v1.3.0
   Compiling lazy_static v1.4.0
   Compiling heck v0.4.0
   Compiling hashbrown v0.12.3
   Compiling serde_derive v1.0.150
   Compiling overload v0.1.1
   Compiling anyhow v1.0.66
   Compiling os_str_bytes v6.4.1
   Compiling strsim v0.10.0
   Compiling textwrap v0.16.0
   Compiling serde v1.0.150
   Compiling termcolor v1.1.3
   Compiling byteorder v1.4.3
   Compiling bitflags v1.3.2
   Compiling serde_json v1.0.89
   Compiling ryu v1.0.11
   Compiling tracing-core v0.1.30
   Compiling thread_local v1.1.4
   Compiling slab v0.4.7
   Compiling tokio v1.23.0
   Compiling num-traits v0.2.15
   Compiling indexmap v1.9.2
   Compiling proc-macro-error-attr v1.0.4
   Compiling proc-macro-error v1.0.4
   Compiling rustls-pemfile v1.0.1
   Compiling pem v1.1.0
   Compiling ring v0.16.20
   Compiling tinyvec v1.6.0
   Compiling sharded-slab v0.1.4
   Compiling regex-automata v0.1.10
   Compiling regex v1.7.0
   Compiling nu-ansi-term v0.46.0
   Compiling clap_lex v0.2.4
   Compiling rustls-native-certs v0.6.2
   Compiling matchers v0.1.0
   Compiling getrandom v0.2.8
   Compiling time v0.3.17
   Compiling socket2 v0.4.7
   Compiling num_cpus v1.14.0
   Compiling signal-hook-registry v1.4.0
   Compiling atty v0.2.14
   Compiling mio v0.8.5
   Compiling rand_core v0.6.4
   Compiling yasna v0.5.0
   Compiling hdrhistogram v7.5.2
   Compiling rand_chacha v0.3.1
   Compiling sct v0.7.0
   Compiling webpki v0.22.0
   Compiling rcgen v0.10.0
   Compiling rand v0.8.5
   Compiling tracing-attributes v0.1.23
   Compiling thiserror-impl v1.0.37
   Compiling tokio-macros v1.8.2
   Compiling clap_derive v3.2.18
   Compiling tracing v0.1.37
   Compiling clap v3.2.23
   Compiling tracing-subscriber v0.3.16
   Compiling quinn-proto v0.9.2 (/home/build/quinn-0.9.3/quinn-proto)
   Compiling quinn-udp v0.3.2 (/home/build/quinn-0.9.3/quinn-udp)
error[E0425]: cannot find value `IP_RECVTOS` in crate `libc`
    --> quinn-udp/src/unix.rs:94:75
     |
94   |         if let Err(err) = set_socket_option(&*io, libc::IPPROTO_IP, libc::IP_RECVTOS, OPTION_ON) {
     |                                                                           ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVIF`
     |
    ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/netbsdlike/openbsd/mod.rs:1068:1
     |
1068 | pub const IP_RECVIF: ::c_int = 30;
     | ---------------------------------- similarly named constant `IP_RECVIF` defined here

error[E0412]: cannot find type `mmsghdr` in crate `libc`
   --> quinn-udp/src/unix.rs:166:26
    |
166 |       let mut msgs: [libc::mmsghdr; BATCH_SIZE] = unsafe { mem::zeroed() };
    |                            ^^^^^^^ help: a struct with a similar name exists: `cmsghdr`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:10:1
    |
10  | / s! {
11  | |     pub struct sockaddr {
12  | |         pub sa_len: u8,
13  | |         pub sa_family: sa_family_t,
...   |
124 | |     }
125 | | }
    | |_- similarly named struct `cmsghdr` defined here

error[E0425]: cannot find function `sendmmsg` in crate `libc`
   --> quinn-udp/src/unix.rs:197:32
    |
197 |         let n = unsafe { libc::sendmmsg(io.as_raw_fd(), msgs.as_mut_ptr(), num_transmits as _, 0) };
    |                                ^^^^^^^^ help: a function with a similar name exists: `sendmsg`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:743:5
    |
743 |     pub fn sendmsg(fd: ::c_int, msg: *const ::msghdr, flags: ::c_int) -> ::ssize_t;
    |     ------------------------------------------------------------------------------- similarly named function `sendmsg` defined here

error[E0412]: cannot find type `mmsghdr` in crate `libc`
   --> quinn-udp/src/unix.rs:296:50
    |
296 |       let mut hdrs = unsafe { mem::zeroed::<[libc::mmsghdr; BATCH_SIZE]>() };
    |                                                    ^^^^^^^ help: a struct with a similar name exists: `cmsghdr`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:10:1
    |
10  | / s! {
11  | |     pub struct sockaddr {
12  | |         pub sa_len: u8,
13  | |         pub sa_family: sa_family_t,
...   |
124 | |     }
125 | | }
    | |_- similarly named struct `cmsghdr` defined here

error[E0425]: cannot find function `recvmmsg` in crate `libc`
   --> quinn-udp/src/unix.rs:308:19
    |
308 |             libc::recvmmsg(
    |                   ^^^^^^^^ help: a function with a similar name exists: `recvmsg`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:748:5
    |
748 |     pub fn recvmsg(fd: ::c_int, msg: *mut ::msghdr, flags: ::c_int) -> ::ssize_t;
    |     ----------------------------------------------------------------------------- similarly named function `recvmsg` defined here

error[E0531]: cannot find unit struct, unit variant or constant `IP_RECVTOS` in crate `libc`
    --> quinn-udp/src/unix.rs:472:73
     |
472  |             (libc::IPPROTO_IP, libc::IP_TOS) | (libc::IPPROTO_IP, libc::IP_RECVTOS) => unsafe {
     |                                                                         ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVIF`
     |
    ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/netbsdlike/openbsd/mod.rs:1068:1
     |
1068 | pub const IP_RECVIF: ::c_int = 30;
     | ---------------------------------- similarly named constant `IP_RECVIF` defined here

Some errors have detailed explanations: E0412, E0425, E0531.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `quinn-udp` due to 6 previous errors
warning: build failed, waiting for other jobs to finish...
sur#
sur# cargo build
   Compiling quinn-udp v0.3.2 (/home/build/quinn-0.9.3/quinn-udp)
error[E0425]: cannot find value `IP_RECVTOS` in crate `libc`
    --> quinn-udp/src/unix.rs:94:75
     |
94   |         if let Err(err) = set_socket_option(&*io, libc::IPPROTO_IP, libc::IP_RECVTOS, OPTION_ON) {
     |                                                                           ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVIF`
     |
    ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/netbsdlike/openbsd/mod.rs:1068:1
     |
1068 | pub const IP_RECVIF: ::c_int = 30;
     | ---------------------------------- similarly named constant `IP_RECVIF` defined here

error[E0412]: cannot find type `mmsghdr` in crate `libc`
   --> quinn-udp/src/unix.rs:166:26
    |
166 |       let mut msgs: [libc::mmsghdr; BATCH_SIZE] = unsafe { mem::zeroed() };
    |                            ^^^^^^^ help: a struct with a similar name exists: `cmsghdr`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:10:1
    |
10  | / s! {
11  | |     pub struct sockaddr {
12  | |         pub sa_len: u8,
13  | |         pub sa_family: sa_family_t,
...   |
124 | |     }
125 | | }
    | |_- similarly named struct `cmsghdr` defined here

error[E0425]: cannot find function `sendmmsg` in crate `libc`
   --> quinn-udp/src/unix.rs:197:32
    |
197 |         let n = unsafe { libc::sendmmsg(io.as_raw_fd(), msgs.as_mut_ptr(), num_transmits as _, 0) };
    |                                ^^^^^^^^ help: a function with a similar name exists: `sendmsg`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:743:5
    |
743 |     pub fn sendmsg(fd: ::c_int, msg: *const ::msghdr, flags: ::c_int) -> ::ssize_t;
    |     ------------------------------------------------------------------------------- similarly named function `sendmsg` defined here

error[E0412]: cannot find type `mmsghdr` in crate `libc`
   --> quinn-udp/src/unix.rs:296:50
    |
296 |       let mut hdrs = unsafe { mem::zeroed::<[libc::mmsghdr; BATCH_SIZE]>() };
    |                                                    ^^^^^^^ help: a struct with a similar name exists: `cmsghdr`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:10:1
    |
10  | / s! {
11  | |     pub struct sockaddr {
12  | |         pub sa_len: u8,
13  | |         pub sa_family: sa_family_t,
...   |
124 | |     }
125 | | }
    | |_- similarly named struct `cmsghdr` defined here

error[E0425]: cannot find function `recvmmsg` in crate `libc`
   --> quinn-udp/src/unix.rs:308:19
    |
308 |             libc::recvmmsg(
    |                   ^^^^^^^^ help: a function with a similar name exists: `recvmsg`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:748:5
    |
748 |     pub fn recvmsg(fd: ::c_int, msg: *mut ::msghdr, flags: ::c_int) -> ::ssize_t;
    |     ----------------------------------------------------------------------------- similarly named function `recvmsg` defined here

error[E0531]: cannot find unit struct, unit variant or constant `IP_RECVTOS` in crate `libc`
    --> quinn-udp/src/unix.rs:472:73
     |
472  |             (libc::IPPROTO_IP, libc::IP_TOS) | (libc::IPPROTO_IP, libc::IP_RECVTOS) => unsafe {
     |                                                                         ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVIF`
     |
    ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/netbsdlike/openbsd/mod.rs:1068:1
     |
1068 | pub const IP_RECVIF: ::c_int = 30;
     | ---------------------------------- similarly named constant `IP_RECVIF` defined here

Some errors have detailed explanations: E0412, E0425, E0531.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `quinn-udp` due to 6 previous errors
sur#
sur# cargo build
   Compiling quinn-udp v0.3.2 (/home/build/quinn-0.9.3/quinn-udp)
error[E0425]: cannot find value `IP_RECVTOS` in crate `libc`
    --> quinn-udp/src/unix.rs:94:75
     |
94   |         if let Err(err) = set_socket_option(&*io, libc::IPPROTO_IP, libc::IP_RECVTOS, OPTION_ON) {
     |                                                                           ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVIF`
     |
    ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/netbsdlike/openbsd/mod.rs:1068:1
     |
1068 | pub const IP_RECVIF: ::c_int = 30;
     | ---------------------------------- similarly named constant `IP_RECVIF` defined here

error[E0412]: cannot find type `mmsghdr` in crate `libc`
   --> quinn-udp/src/unix.rs:166:26
    |
166 |       let mut msgs: [libc::mmsghdr; BATCH_SIZE] = unsafe { mem::zeroed() };
    |                            ^^^^^^^ help: a struct with a similar name exists: `cmsghdr`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:10:1
    |
10  | / s! {
11  | |     pub struct sockaddr {
12  | |         pub sa_len: u8,
13  | |         pub sa_family: sa_family_t,
...   |
124 | |     }
125 | | }
    | |_- similarly named struct `cmsghdr` defined here

error[E0425]: cannot find function `sendmmsg` in crate `libc`
   --> quinn-udp/src/unix.rs:197:32
    |
197 |         let n = unsafe { libc::sendmmsg(io.as_raw_fd(), msgs.as_mut_ptr(), num_transmits as _, 0) };
    |                                ^^^^^^^^ help: a function with a similar name exists: `sendmsg`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:743:5
    |
743 |     pub fn sendmsg(fd: ::c_int, msg: *const ::msghdr, flags: ::c_int) -> ::ssize_t;
    |     ------------------------------------------------------------------------------- similarly named function `sendmsg` defined here

error[E0412]: cannot find type `mmsghdr` in crate `libc`
   --> quinn-udp/src/unix.rs:296:50
    |
296 |       let mut hdrs = unsafe { mem::zeroed::<[libc::mmsghdr; BATCH_SIZE]>() };
    |                                                    ^^^^^^^ help: a struct with a similar name exists: `cmsghdr`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:10:1
    |
10  | / s! {
11  | |     pub struct sockaddr {
12  | |         pub sa_len: u8,
13  | |         pub sa_family: sa_family_t,
...   |
124 | |     }
125 | | }
    | |_- similarly named struct `cmsghdr` defined here

error[E0425]: cannot find function `recvmmsg` in crate `libc`
   --> quinn-udp/src/unix.rs:308:19
    |
308 |             libc::recvmmsg(
    |                   ^^^^^^^^ help: a function with a similar name exists: `recvmsg`
    |
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/mod.rs:748:5
    |
748 |     pub fn recvmsg(fd: ::c_int, msg: *mut ::msghdr, flags: ::c_int) -> ::ssize_t;
    |     ----------------------------------------------------------------------------- similarly named function `recvmsg` defined here

error[E0531]: cannot find unit struct, unit variant or constant `IP_RECVTOS` in crate `libc`
    --> quinn-udp/src/unix.rs:472:73
     |
472  |             (libc::IPPROTO_IP, libc::IP_TOS) | (libc::IPPROTO_IP, libc::IP_RECVTOS) => unsafe {
     |                                                                         ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVIF`
     |
    ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.138/src/unix/bsd/netbsdlike/openbsd/mod.rs:1068:1
     |
1068 | pub const IP_RECVIF: ::c_int = 30;
     | ---------------------------------- similarly named constant `IP_RECVIF` defined here

Some errors have detailed explanations: E0412, E0425, E0531.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `quinn-udp` due to 6 previous errors

```

###

djc commented:
Let's discuss OpenBSD in #1469 (@lucifer9 any relevant expertise you have would be much appreciated!).
